### DISCUSSION TITLE
unsupported layers

### DISCUSSION THREAD

**Discussion by abdullah35mts** [fastmachinelearning/hls4ml#648] [General]
I have a simple model with conv1D  with pad ,adaptiveavgpool1d and a squeeze function , when i tried to convert the model form onnx to hls using hls4ml.converter_onnx_to_hls , it gave me unsupported layer 'pad' error , i remove the padding layer then it gave me unsupported layer 'squeeze' then I tried using skip layer by adding squeeze to skip layer list and again i encounter the following error
![1](https://user-images.githubusercontent.com/45236771/188827661-474cf441-bb98-4351-a62b-e03d3a8c175c.png)
![2](https://user-images.githubusercontent.com/45236771/188827666-121818ad-d28a-4b5e-a173-2e191fce7418.png)
My question is , do we have pad and squeeze function in hls4ml . If yes then why i am getting the error of not supported layers and when i go through the onnx_to_hls.py i can see the pad function implemented there.
Any help will be much appertained

**Comment by jmitrevs** [comment]
The onnx front end is undergoing quite a few revisions, with the new version in the ingest-qonnx-master branch. The new version depends on some preprocessing, however, from the https://github.com/fastmachinelearning/qonnx toolkit, in particular, one needs to do some cleaning (https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/util/cleanup.py) and convert to channels last (https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/util/to_channels_last.py). Those exist as standalone executibles when you install qunnox, too.

One thing I noticed is that your network has a Gemm. That's not supported in the new parsing, and one should use the https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/transformation/gemm_to_matmul.py transfromation to convert it to MatMul and Add. Currently I believe that only works if your nodes are single precision, not double precision. (We will make it more universal in the future.)

**Reply by vloncar** [reply]
`Gemm` is [supported](https://github.com/fastmachinelearning/hls4ml/blob/main/hls4ml/converters/onnx/core.py#L5). In fact, we should prefer Gemm over MatMul+Add as it is far easier to parse and maps directly to `Dense`. ONNX itself has an [optimizer](https://github.com/onnx/optimizer/blob/master/onnxoptimizer/passes/fuse_matmul_add_bias_into_gemm.h#L29) to fuse MatMul+Add into Gemm.

**Reply by jmitrevs** [reply]
It is supported by the old parser, but not the new. We decided to just support one style in the new and use the common optimizer to go from one to the other.

**Reply by vloncar** [reply]
Just to understand this... It has been decided to discard support for a part of ONNX spec because that has been the preference on another project not related to hls4ml, and hls4ml will accept this, reinvent the wheel again with the optimizer and take the maintenance cost of it going forward? :man_facepalming:

**Reply by jmitrevs** [reply]
In a sense, it's being moved to the common front end. Calling the common optimizers (including going to channels last) will eventually be automated. It was to not have to manually add support for Gemm with Quant nodes and scale propagations if the optimizer that makes that redundant exists in the common front end. It's for the same reason that there is not support for channels first with QONNX. We offload that to the common front end.

**Comment by abdullah35mts** [comment]
I have tried to make change the channels to last but i get this error 
<img width="582" alt="Screenshot 2022-09-21 at 20 08 54" src="https://user-images.githubusercontent.com/45236771/191579587-6ce54611-29ae-482e-9f97-5d8e98969208.png">
It only works for 3d or 4d tensor ,but my model has conv1d . Is there a way i can still use it.
[clean_model.onnx.zip](https://github.com/fastmachinelearning/hls4ml/files/9619311/clean_model.onnx.zip)
I have attached my model along as well to reproduce the error.
Any help will be much appreciated
